### ISSUE TITLE
Failing unit tests with libmxml 2.10

### ISSUE THREAD

**Issue by quantumsteve** [nexusformat/code#443]
The following unit tests segfault when building with libmxml v2.10. The same tests pass with libmxml v2.9.

```
$ ctest
Test project /Users/svh/Documents/nexusformat/build
    Start 1: NAPI-C-HDF5-test
1/7 Test #1: NAPI-C-HDF5-test .................   Passed    0.06 sec
    Start 2: NAPI-C-HDF5-attra-test
2/7 Test #2: NAPI-C-HDF5-attra-test ...........   Passed    0.01 sec
    Start 3: NAPI-C-MXML-test
3/7 Test #3: NAPI-C-MXML-test .................***Exception: SegFault  0.45 sec
    Start 4: NAPI-C-MXML-TABLE-test
4/7 Test #4: NAPI-C-MXML-TABLE-test ...........***Exception: SegFault  0.25 sec
    Start 5: NAPI-C-MXML-attra-test
5/7 Test #5: NAPI-C-MXML-attra-test ...........***Exception: SegFault  0.24 sec
    Start 6: NAPI-C-leak-test-1
6/7 Test #6: NAPI-C-leak-test-1 ...............   Passed    0.63 sec
    Start 7: NAPI-C-test-nxunlimited
7/7 Test #7: NAPI-C-test-nxunlimited ..........***Exception: SegFault  0.28 sec

43% tests passed, 4 tests failed out of 7

Total Test time (real) =   1.94 sec

The following tests FAILED:
	  3 - NAPI-C-MXML-test (SEGFAULT)
	  4 - NAPI-C-MXML-TABLE-test (SEGFAULT)
	  5 - NAPI-C-MXML-attra-test (SEGFAULT)
	  7 - NAPI-C-test-nxunlimited (SEGFAULT)
Errors while running CTest
$ 
```

I ran `napi_attra_test_xml` through lldb and it stops at line 549 in `nxio.c`. Specifically, `node->parent = NULL`

```
Process 47139 stopped
* thread #1: tid = 0xd1836, 0x0000000100020863 libNeXus.1.dylib`nexusWriteCallback(node=0x0000000100404bc0) + 99 at nxio.c:549, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x18)
    frame #0: 0x0000000100020863 libNeXus.1.dylib`nexusWriteCallback(node=0x0000000100404bc0) + 99 at nxio.c:549
   546 	  /* this is set by nxconvert when making a definiton */
   547 	  is_definition = (getenv("NX_IS_DEFINITION") != NULL);
   548 	
-> 549 	  if (!strcmp(node->parent->parent->value.element.name, DATA_NODE_NAME))
   550 	  {
   551 	    table_style = 1;
   552 	  }
```

The backtrace is
```
(lldb) bt
* thread #1: tid = 0xd1836, 0x0000000100020863 libNeXus.1.dylib`nexusWriteCallback(node=0x0000000100404bc0) + 99 at nxio.c:549, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x18)
  * frame #0: 0x0000000100020863 libNeXus.1.dylib`nexusWriteCallback(node=0x0000000100404bc0) + 99 at nxio.c:549
    frame #1: 0x0000000100034933 libmxml.dylib`mxml_write_node(node=<unavailable>, p=<unavailable>, cb=<unavailable>, col=0, putc_cb=<unavailable>, global=<unavailable>) + 603 at mxml-file.c:2928
    frame #2: 0x0000000100034fbf libmxml.dylib`mxmlSaveFile(node=<unavailable>, fp=0x00007fff78383070, cb=<unavailable>) + 48 at mxml-file.c:332
    frame #3: 0x0000000100018ab8 libNeXus.1.dylib`NXXclose(fid=0x00007fff5fbff6c8) + 216 at nxxml.c:227
    frame #4: 0x0000000100008e07 libNeXus.1.dylib`nxiclose_(fid=0x00007fff5fbff7b8) + 103 at napi.c:700
    frame #5: 0x0000000100002b4b napi_attra_test_xml`main(argc=1, argv=0x00007fff5fbffa50) + 1131 at napi_attra_test.c:147
    frame #6: 0x00007fff8ae275c9 libdyld.dylib`start + 1
```

**Comment by zjttoefs** [comment]
@eugenwintersberger do you have time to have a look? Otherwise I could try.

**Comment by eugenwintersberger** [comment]
I am currently on vacation in Luxor egypt. So I can deal with thus issue earliest in two weeks.

**Comment by zjttoefs** [comment]
Looks like Luxor has internet, so what’s stopping you?

;-)

Have fun!


I’ll see how much success I have setting up a test in limited time.

**Comment by zjttoefs** [comment]
If nothing else, I've at least confirmed the problem now.

**Comment by zjttoefs** [comment]
Very difficult for me to find out what's going on there. I'm not familiar how things map into XML or the MXML data structures. Would be better if someone who understands that already has a look.

**Comment by mkoennecke** [comment]
Hi,

I tracked this down. It is actually what I consider a bug in mxm-2.10. 

The problem is in line 2928 of file mxml-file.c in the mxml-2.10 source code. 
In the distributed version the line reads:

    if ((data = (*global->custom_save_cb)(node)) == NULL)

but it must read:

    if ((data = (*global->custom_save_cb)(current)) == NULL)

With that change the NeXus XML API test started working again. From the logic in that function it is obvious that it must be the node we are writing, current, rather then the one the function got called with, node, which is required as a parameter here.

I created an issue in the mxml issue tracker for this. Until I get feedback from the maintainer of mxml I consider this as dormant. Not much I can do if my callback is called with wrong parameters...
